spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/Spark3ZOrderStrategy.java

```diff
@@ -0,0 +1,302 @@
+  private static final int STRING_KEY_LENGTH = 60;
+
+  private final List<String> zOrderColNames;
+  private final FileScanTaskSetManager manager = FileScanTaskSetManager.get();
+  private final FileRewriteCoordinator rewriteCoordinator = FileRewriteCoordinator.get();
+
+  public Spark3ZOrderStrategy(Table table, SparkSession spark, List<String> zOrderColNames) {
+    super(table, spark);
+
+    Stream<String> identityPartitionColumns = table.spec().fields().stream()
+        .filter(f -> f.transform().isIdentity())
+        .map(PartitionField::name);
+    List<String> partZOrderCols = identityPartitionColumns
+        .filter(zOrderColNames::contains)
+        .collect(Collectors.toList());
+    Preconditions.checkArgument(
+        partZOrderCols.isEmpty(),
+        "Cannot ZOrder on an Identity partition column as these values are constant within a partition, " +
+            "ZOrdering requested on %s",
+        partZOrderCols);
+
+    this.zOrderColNames = zOrderColNames;
+  }
+
+  @Override
+  public String name() {
+    return "Z-ORDER";
+  }
+
+  @Override
+  protected void validateOptions() {
+    // TODO implement Zorder Strategy in API Module
+    return;
+  }
+
+  @Override
+  public Set<DataFile> rewriteFiles(List<FileScanTask> filesToRewrite) {
+    String groupID = UUID.randomUUID().toString();
+    boolean requiresRepartition = !filesToRewrite.get(0).spec().equals(table().spec());
+
+    SortOrder[] ordering;
+    if (requiresRepartition) {
+      ordering = SparkDistributionAndOrderingUtil.convert(SortOrderUtil.buildSortOrder(table(), sortOrder()));
+    } else {
+      ordering = SparkDistributionAndOrderingUtil.convert(sortOrder());
+    }
+
+    Distribution distribution = Distributions.ordered(ordering);
+
+    try {
+      manager.stageTasks(table(), groupID, filesToRewrite);
+
+      // Disable Adaptive Query Execution as this may change the output partitioning of our write
+      SparkSession cloneSession = spark().cloneSession();
+      cloneSession.conf().set(SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(), false);
+
+      // Reset Shuffle Partitions for our sort
+      long numOutputFiles = numOutputFiles((long) (inputFileSize(filesToRewrite) * sizeEstimateMultiple()));
+      cloneSession.conf().set(SQLConf.SHUFFLE_PARTITIONS().key(), Math.max(1, numOutputFiles));
+
+      Dataset<Row> scanDF = cloneSession.read().format("iceberg")
+          .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
+          .load(table().name());
+
+      Column[] originalColumns = Arrays.stream(scanDF.schema().names())
+          .map(n -> functions.col(n))
+          .toArray(Column[]::new);
+
+      List<StructField> zOrderColumns = zOrderColNames.stream()
+          .map(scanDF.schema()::apply)
+          .collect(Collectors.toList());
+
+      Column zvalueArray = functions.array(zOrderColumns.stream().map(colStruct ->
+          SparkZOrder.sortedLexicographically(functions.col(colStruct.name()), colStruct.dataType())
```

Comment:
I'm not sure that `sortedLexicographically` makes sense here. It looks like this is converting values to byte arrays that have the same sort, but there's no real sorting going on here, right?